datafusion/expr/src/udf.rs

```diff
@@ -307,12 +308,17 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
         &self,
         args: &[Expr],
         schema: &dyn ExprSchema,
+        arg_types: &[DataType],
     ) -> Result<DataType> {
-        let arg_types = args
-            .iter()
-            .map(|arg| arg.get_type(schema))
-            .collect::<Result<Vec<_>>>()?;
-        self.return_type(&arg_types)
+        if arg_types.is_empty() {
```

Comment:
I wonder if `arg_types` is empty, would here also get the empty `arg_types` since this param is calculated in the same way before? 🤔

Reply:
if `arg_types` is not provided, we can get them from `args` and `schema`, and it should not be empty.

Reply:
When will `arg_types` not be provided?
I think in the current PR, we will pass always pass `arg_types` to the default implementation of `return_type_from_exprs`, thus it may be alway non-empty?